apps/src/sites/studio/pages/levels/show.js

```diff
@@ -59,12 +60,18 @@ function initPage() {
 
   if (hasScriptData('script[data-aitutorleveldata]')) {
     const aiTutorLevelData = getScriptData('aitutorleveldata');
+    const {id, type, hasValidation} = aiTutorLevelData;
+    const level = {
+      id: id,
+      type: type,
+      hasValidation: hasValidation,
+    };
```

Comment:
❓ Is there a way to call setLevel with the aitutorleveldata in levels/show.js rather than passing it down to all of the children and grandchildren components? Basically something like [this](https://github.com/code-dot-org/code-dot-org/blob/75443933100d3d3e7bd45da546588379c7d18cc7/apps/src/sites/studio/pages/levels/show.js#L84) but I wasn't sure what syntax to use with TypeScript and the way aiTutorRedux is set up.

Member
@[sanchitmalhotra126](https://github.com/sanchitmalhotra126) sanchitmalhotra126 [4 days ago](https://github.com/code-dot-org/code-dot-org/pull/56218#discussion_r1476728389)
I think we generally have a preference to avoid using redux code outside of React files (even though we have plenty of exceptions in older code 😅). I'm not as familiar with this feature, but where is the level information originating from? Is this tool meant to be used with any lab, or just Java Lab?

Reply:
The level information is coming in from the view at [levels/show.haml.html](https://github.com/code-dot-org/code-dot-org/blob/344ff55a8ecc7decf5ab0c4151d04c0eff067969/dashboard/app/views/levels/show.html.haml#L19) and passed onto to show.js 

The short term answer is the tool is meant to used in the context of CSA levels. The eventual potential long term answer is that the tool is meant to be used much more broadly across labs, level types and maybe even outside of levels. 

I don't want to use an old/anti-pattern was just curious if there's a current preference to find a way to put data shared by multiple components into the store and accessing it from there or if it's ok, if a tiny bit cumbersome, to pass it around as needed. 🤷‍♀️

Reply:
I think either way is fine, for lab2 levels we load the level data client side and immediately store in redux, but when it comes from the server the pattern is a little messier. This file is already adding things to redux, IMO it would be fine to set the level data here too. See line 32 in this file:
`getStore().dispatch(setCodeReviewEnabledForLevel(codeReviewEnabled));`

Reply:
Another option could be to use React [context](https://react.dev/learn/passing-data-deeply-with-context) to avoid passing props through multiple layers of components, but to Molly's point, looks like we're already passing server data into redux here, so that works too.